### ISSUE TITLE
requirements.txt pins almost nothing: venv and container diverged by 120 packages (18 major)

### ISSUE THREAD

**Issue by attevon-admin** [davidamacey/OpenTranscribe#492]
## Summary

`backend/requirements.txt` pins almost nothing: of 68 specs, **4** are exact `==`, **61** are
`>=` floors, and 2 are unversioned. A floor is a version number but not a pin — `nltk>=3.9.4`
permits 3.9.4 *or* 3.10.3 — so two installs performed at different times legitimately resolve
to different versions.

They have. Measured between the host venv (where `./scripts/run-integration-tests.sh`, the
pre-merge gate, runs) and the running backend container:

- **120 packages differ**, of which **18 differ at a MAJOR version**.
- Every `==`-pinned package matched **exactly** (`torch==2.11.0`, `onnxruntime-gpu==1.28.0`).
- Every drifted package sat **exactly on its floor** in the venv and at latest in the container.

Major-version divergences observed:

| package | venv | container |
|---|---|---|
| starlette | 0.48.0 | **1.6.0** |
| openai | 2.44.0 | **3.2.0** |
| pandas | 2.2.3 | **3.0.5** |
| protobuf | 6.33.6 | **7.35.1** |
| assemblyai | 0.64.25 | **1.0.0** |
| av | 15.1.0 | **18.1.0** |
| pillow | 11.3.0 | 12.3.0 |
| rich | 14.2.0 | 15.0.0 |
| pycparser | 2.23 | 3.0 |
| attrs, certifi, fsspec, packaging, pytz, regex, setuptools, tzdata | (older) | (newer) |

Plus 102 minor/patch differences including `sqlalchemy`, `alembic`, `redis`, `uvicorn`,
`transformers`, `huggingface-hub`, `numpy`, `scikit-learn`, `spacy`.

4 packages exist **only** in the container: `httpcore2`, `httpx2`, `narwhals`, `truststore`.

## Why this matters

**The pre-merge gate does not test what ships.** `starlette` is the clearest example: the gate
validated 0.48 while production runs 1.6 — and Starlette's threadpool behaviour is precisely
what issue #485 turned on.

This is not hypothetical. It already caused a production outage-class defect to pass the gate:
**#491 / the NLTK `pathsec` breakage.** `nltk>=3.9.4` resolved to 3.9.4 in the venv and 3.10.3
in the image. 3.10.3 added a hardening layer that refuses multiply-linked files, so
`split_sentences_nltk` raised on **every** call in the backend and both transcription
workers — while the host test suite passed, because it was running the older NLTK. The gate
was structurally incapable of seeing it.

## Suggested fix

Standard two-file layout, used by **both** the Dockerfile and the venv:

1. `requirements.txt` stays the human-edited **direct** dependency list.
2. A generated `requirements.lock.txt` holds exact `==` pins for the full resolved tree
   (the known-good set the image currently runs is a reasonable seed).
3. `Dockerfile.prod` / `.lite` / `.blackwell` install from the lock.
4. The venv bootstrap in `backend/CLAUDE.md` installs from the same lock.
5. Regenerating the lock becomes a deliberate, reviewable commit — which is also what makes a
   dependency bump visible in a PR diff instead of invisible in a rebuild.

Note `requirements-nodeps.txt` (whisperx / faster-whisper / gliner, installed `--no-deps`)
needs the same treatment, and `requirements-ci.txt` must not drift from the lock either.

### Interim state

The venv has been aligned by hand to the container's exact versions (118 packages), so the
gate currently compares like with like. **That alignment is not reproducible** — a fresh
`pip install -r requirements.txt` will drift again the moment upstream publishes. This issue
is the durable fix.

## Verification

After locking: build the image, create a fresh venv from the lock, and diff
`pip list --format=freeze` between the two. The correct result is an empty diff apart from
`pip`/`setuptools`. That diff is worth wiring into CI — it is a cheap check that would have
caught every item above.

Refs #491

**Comment by attevon-admin** [comment]
## Two traps found while scoping this — a naive `pip freeze` lock will not work

Recording these so whoever picks this up doesn't discover them mid-implementation.

**1. `pyannote.audio` is pinned to a MOVING BRANCH, not a commit.**

```
pyannote.audio @ git+https://github.com/davidamacey/pyannote-audio.git@gpu-optimizations
```

`gpu-optimizations` advances, so this line installs something different over time by design.
The currently-installed build is `4.0.5.dev16+ga3f38afb8` — i.e. commit `a3f38afb8`. This is a
second, independent reproducibility hole in the same file, and arguably a sharper one than the
floors: nothing in the repo records which fork commit a given image was built from. Pin the SHA
(`...@a3f38afb8`) and bump it deliberately.

**2. `torch`/`torchaudio` come from a custom index.**

`requirements.txt:71` carries `--extra-index-url https://download.pytorch.org/whl/cu128`, and the
resolved builds are `2.11.0+cu128`. A lockfile must preserve both the local version suffix and the
index, or the build silently falls back to CPU wheels from PyPI.

**3. `requirements-nodeps.txt`** (whisperx, faster-whisper, gliner) installs with `--no-deps` and
must stay a separate step — see that file's header for why.

So the lock cannot be a flat `pip freeze` of the image; it needs the three cases above handled
explicitly, which is most of the work in this issue.

## Interim state after PR for #485/#486/#487

The venv was aligned **by hand** to the container's exact versions (118 packages plus 4 that
existed only in the container: `httpcore2`, `httpx2`, `narwhals`, `truststore`). `pip list
--format=freeze` now differs only by `pip`/`setuptools`.

That alignment immediately surfaced a real failure the drift had been hiding: `openai` 3.2.0
imports `httpx2`, which the venv did not have, so 10 tests in `test_openai_provider.py` and
`test_provider_sdk_compat.py::test_client_importable` failed. All pass once the container-only
packages are installed. **This is exactly the class of defect the floors conceal** — the gate had
been validating `openai` 2.44 while production runs 3.2.

⚠️ **The alignment is not reproducible and will drift again** on the next `pip install -r
requirements.txt`. Only the `nltk>=3.10.3,<4` floor raise landed in the repo (see #491).

**Comment by attevon-admin** [comment]
## Fixed — `dcf4a3a9` (supersedes `636af0a8`)

Fixed on branch `fix/v050-gap-closure`. **Not yet merged** — this issue closes when the PR lands.

⚠️ **The approach in `636af0a8` was reverted.** That commit added a generated `requirements.lock.txt` plus `scripts/lock-backend-deps.sh` — a bespoke two-artefact setup. It is replaced by the ordinary thing other repos do: **one fully pinned requirements file per environment**, installed by both its container and the venv, so "what a developer installs" and "what ships" are the same text rather than two artefacts that can disagree.

All **209 specs across five files** are now exact pins.

**Measured after, venv vs the rebuilt dev container: 279 shared, 0 mismatched, 0 image-only** (from 120 apart, 18 at a major version).

Four defects the pinning exposed:

- **`Dockerfile.blackwell` could not `import transformers` at all** (#500) — `huggingface_hub==0.23.5` against transformers' `>=0.34.0`, enforced at import. `pip` reports the conflict as a *warning* and exits 0, so the image built green.
- **The spaCy model was installed by a `RUN` line in `Dockerfile.prod` only** — so `Dockerfile.blackwell` shipped PII redaction with no pipeline to tokenize with, and a developer's venv never had it. Now pinned by release URL.
- **`pyannote.audio` was installed from a moving branch** — a line that looks specific and is not.
- **`starlette` was pinned nowhere**, only ranged by fastapi — and its threadpool behaviour is exactly what #485 turns on.

Five transitive packages are pinned too, with the reason inline: after every direct dependency was pinned and both sides rebuilt, they were the only application packages still disagreeing.

**Guards:** `tests/unit/test_requirements_pinned.py` (13 of its tests verified red against HEAD) and `scripts/check-dependency-parity.sh`, now phase 3b of the pre-merge gate.

⚠️ Two of those guards were themselves wrong first, and both are recorded in the module: the git test accepted `@gpu-optimizations` (a branch *is* a revision), so it passed against the exact defect it was written to catch; and the Dockerfile-override detector read comments as code.
